core/state/access_events.go

```diff
@@ -152,33 +154,51 @@ func (ae *AccessEvents) AddTxDestination(addr common.Address, sendsValue bool) {
 }
 
 // SlotGas returns the amount of gas to be charged for a cold storage access.
-func (ae *AccessEvents) SlotGas(addr common.Address, slot common.Hash, isWrite bool) uint64 {
+func (ae *AccessEvents) SlotGas(addr common.Address, slot common.Hash, isWrite bool) (*multigas.MultiGas, error) {
 	treeIndex, subIndex := utils.StorageIndex(slot.Bytes())
-	return ae.touchAddressAndChargeGas(addr, *treeIndex, subIndex, isWrite)
+	return ae.touchAddressAndChargeMultigas(addr, *treeIndex, subIndex, isWrite)
 }
 
-// touchAddressAndChargeGas adds any missing access event to the access event list, and returns the cold
+// touchAddressAndChargeMultigas adds any missing access event to the access event list, and returns the cold
 // access cost to be charged, if need be.
-func (ae *AccessEvents) touchAddressAndChargeGas(addr common.Address, treeIndex uint256.Int, subIndex byte, isWrite bool) uint64 {
+func (ae *AccessEvents) touchAddressAndChargeMultigas(addr common.Address, treeIndex uint256.Int, subIndex byte, isWrite bool) (*multigas.MultiGas, error) {
 	stemRead, selectorRead, stemWrite, selectorWrite, selectorFill := ae.touchAddress(addr, treeIndex, subIndex, isWrite)
 
-	var gas uint64
-	if stemRead {
-		gas += params.WitnessBranchReadCost
+	gas := multigas.ZeroGas()
+	// Reading includes witness calculation (computation), and storage access
+	// The computation cost for the witness calculation is negligible compared to the storage access cost
+	if stemRead && gas.SafeIncrement(multigas.ResourceKindStorageAccess, params.WitnessBranchReadCost) {
+		return nil, fmt.Errorf("failed to increment gas for branch read due to overflow")
```

Comment:
I would not check for overflow in this function because it is impossible to happen. Also, I would prefer to keep the code structure similar to what it was before to make future merges easier.